runtime-sdk/src/state.rs

```diff
@@ -0,0 +1,1538 @@
+use std::{
+    any::Any,
+    cell::RefCell,
+    collections::btree_map::{BTreeMap, Entry},
+    fmt,
+    marker::PhantomData,
+    mem,
+};
+
+use oasis_core_runtime::{common::crypto::hash::Hash, consensus::roothash, storage::mkvs};
+
+use crate::{
+    context::Context,
+    crypto::random::RootRng,
+    event::{Event, EventTag, EventTags},
+    modules::core::Error,
+    storage::{MKVSStore, NestedStore, OverlayStore, Store},
+    types::{address::Address, message::MessageEventHookInvocation, transaction},
+};
+
+/// Execution mode.
+#[derive(Clone, Copy, Default, Debug, PartialEq, Eq)]
+#[repr(u8)]
+pub enum Mode {
+    /// Actually execute transactions during block production.
+    #[default]
+    Execute,
+    /// Check that transactions are valid for local acceptance into the transaction pool.
+    Check,
+    /// Simulate transaction outcomes (e.g. for gas estimation).
+    Simulate,
+    /// Check that transactions are still valid before scheduling.
+    PreSchedule,
+}
+
+const MODE_CHECK: &str = "check";
+const MODE_EXECUTE: &str = "execute";
+const MODE_SIMULATE: &str = "simulate";
+const MODE_PRE_SCHEDULE: &str = "pre_schedule";
+
+impl fmt::Display for Mode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(self.into())
+    }
+}
+
+impl From<&Mode> for &'static str {
+    fn from(m: &Mode) -> Self {
+        match m {
+            Mode::Check => MODE_CHECK,
+            Mode::Execute => MODE_EXECUTE,
+            Mode::Simulate => MODE_SIMULATE,
+            Mode::PreSchedule => MODE_PRE_SCHEDULE,
+        }
+    }
+}
+
+/// Information about the execution environment.
+#[derive(Clone, Default, Debug)]
+pub struct Environment {
+    mode: Mode,
+    tx: Option<TransactionWithMeta>,
+    internal: bool,
+}
+
+impl Environment {
+    /// Execution mode.
+    pub fn mode(&self) -> Mode {
+        self.mode
+    }
+
+    /// Whether the execution mode is such that only checks should be performed.
+    pub fn is_check_only(&self) -> bool {
+        matches!(self.mode, Mode::Check | Mode::PreSchedule)
+    }
+
+    /// Whether the execution mode is `Mode::PreSchedule`.
+    pub fn is_pre_schedule(&self) -> bool {
+        matches!(self.mode, Mode::PreSchedule)
+    }
+
+    /// Whether the execution mode is `Mode::Simulate`.
+    pub fn is_simulation(&self) -> bool {
+        matches!(self.mode, Mode::Simulate)
+    }
+
+    /// Whether the execution mode is `Mode::Execute`.
+    pub fn is_execute(&self) -> bool {
+        matches!(self.mode, Mode::Execute)
+    }
+
+    /// Whether there is an active transaction in the current environment.
+    pub fn is_transaction(&self) -> bool {
+        self.tx.is_some()
+    }
+
+    /// An active transaction's index (order) within the block. If there is no active transaction,
+    /// this method will return zero.
```

Comment:
Should this and other tx-only methods panic if not called from within a transaction environment (like tx_auth_info does)?

Reply:
Have the same question, as the caller cannot distinguish between the right and the default index.

Reply:
Changed, please take a look.